### DISCUSSION TITLE
자료구조 - 트리 - 호현&일구&성은

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#57] [질의 응답]
- B트리와 B+트리의 구조와 그 차이점에 대해 설명해주세요
- 레드-블랙트리에 대해 설명해주세요.
- 레드-블랙트리에 새로운 데이터를 삽입할 때, 내부 동작 방식에 대해 설명해주세요.
- AVL트리에 대해 해당 트리의 삽입, 검색, 삭제의 시간 복잡도를 포함하여 설명해주세요.
- 트리 순회를 위해 사용하는 방법에는 무엇이 있는지 설명해주세요.
- 주어진 트리의 높이와 차수를 말해주세요.
- BST(Binary Search Tree)와 Binary Tree에 대해 설명해주세요.

**Comment by seongeun42** [comment]
### ❓️ B트리와 B+트리의 구조와 그 차이점에 대해 설명해주세요

<br>

### 💬 답변
<blockquote>

![B-tree](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/f58507ca-d651-4450-a366-a6573671f31e)

B트리는 보통 B-트리를 뜻한다. 이진 트리를 확장한 B트리는 하나의 노드에 여러 개의 값(M개)을 가질 수 있고, 두 개 이상의 자식을 가질 수 있다. 최대 M개의 자식을 가지는 B트리를 M차 B트리라고 한다. 

B트리의 가장 중요한 특징은 균형을 유지한다는 것이다. B트리의 각 노드 값은 정렬되어 있고, 중복되지 않으며, 모든 리프 노드는 높이가 같다(즉, 가장 마지막 높이에 모든 리프 노드가 있다). 또한 각 노드는 리프 노드가 아닌 이상 최소 2개 이상의 자식 노드를 가진다. 루트 노드와 리프 노드를 제외한 노드들은 최소 M/2개에서 최대 M개의 자식 노드를 가질 수 있다.

B트리는 정렬 상태를 유지하기 때문에 탐색에 활용하면 좋다. 탐색은 루트 노드에서 시작해 하향식으로 찾는다. 하지만 반대로, 정렬 상태를 유지하기 때문에 새로운 노드의 삽입 혹은 기존 노드 삭제 시 균형을 유지하기 위해 복잡한 연산 과정을 거쳐야 한다.

B+트리는 B트리의 변형된 형태로, 보다 효율적인 삽입, 삭제, 탐색을 위해 고안되었다. B+트리의 노드는 Index 노드와 Data 노드로 나뉘는데, Index 노드는 리프 노드를 제외한 모든 노드이고, Data 노드는 리프 노드를 의미한다. 즉, 실제 데이터는 전부 리프 노드에 저장되고, 나머지는 인덱스(키)이다. 리프 노드들은 연결리스트로 연결되어 선형적 탐색이 가능하며, 범위 탐색에 유리하다. 또한 데이터의 삽입, 삭제가 리프 노드에서만 이루어지기 때문에 더 효율적이다.

B-트리는 모든 노드에 데이터와 키 값을 함께 저장한다. 반면 B+트리는 내부 노드에는 키 값만 저장하고, 리프 노드에 데이터를 저장한다.

B+트리는 DB Index로 많이 사용되는 자료구조 중 하나이다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 레드-블랙트리에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

![Red-black_tree_example svg](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/c4ce4259-8ab3-4a14-9af8-390a84dd8c7e)

레드-블랙 트리는 자가 균형 이진 탐색 트리로서, n개의 원소가 있을 때 최악의 경우에도 $O(logN)$의 시간 복잡도로 삽입, 삭제, 검색을 할 수 있다. 레드-블랙 트리는 자식 노드가 존재하지 않는 경우 NIL이라는 특수 노드를 둔다. 즉, 모든 Leaf 노드는 NIL이다.

레드-블랙 트리는 다음 5가지 조건을 만족한다.

1. 각 노드의 색은 Red 혹은 Black
2. Root 노드는 Black
3. 모든 Leaf 노드(NIL)는 Black
4. Red 노드의 자식 노드들은 전부 Black
    - Red 노드의 부모, 자식은 모두 Black으로, Red가 연속으로 등장X
    - Black은 상관 없다
5. Root 노드에서 시작해 Leaf 노드에 이르기까지 모든 경로에는 동일한 개수의 Black 노드가 존재

마지막 조건을 예를 들어 말하자면, `13 - 8 - 1 - 6 - NIL`에서 13, 1, NIL이 Black으로 해당 경로에는 Black이 3개 존재한다. `13 - 17 - 15 - NIL` 또한 13, 15, NIL이 Black으로 3개 존재하며, 다른 경로들도 마찬가지다. 이 조건을 통해 트리의 높이를 균형 있게 유지함으로써 탐색 시간이 $O(logN)$이 되도록 보장한다.

</blockquote>

[Red-Black Tree 시뮬레이션 사이트](https://www.cs.usfca.edu/~galles/visualization/RedBlack.html)

<br>

**Reply by seongeun42** [reply]
### ❓️ 레드-블랙트리에 새로운 데이터를 삽입할 때, 내부 동작 방식에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

* RB 트리에 새로운 노드는 항상 **Red**로 삽입된다. 
   * 이로인해 4번 조건이 위배되는 케이스가 등장할 수 있다.

![rb1](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/28c72614-b77e-49f7-854b-10f3c144783c)

* 새로 삽입할 노드를 N(New), 부모 노드를 P(Parent), 조상 노드를 G(Grand Parent), 삼촌 노드를 U(Uncle)라고 하자.
   * 삼촌 노드 = 부모 노드 P와 같은 부모 G를 가진 노드
* U가 Black이라면 **Restructuring**
* U가 Red라면 **Recoloring**

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 레드-블랙트리의 Restructuring에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

1. N, P, G를 오름차순으로 정렬한다.
2. 셋 중 중간값을 부모로 만들고 나머지 둘을 자식으로 만든다.
3. 새로 부모가 된 노드를 검은색으로 만들고 나머지 자식들을 빨간색으로 만든다.


* 오름차순 정렬
![rb-2](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/e15af3f7-3936-433c-adbb-4958a1a366f8)

* 중간값을 부모노드로
   * 이때 7은 원래 5의 자식노드이므로 그 관계가 유지됨 
![rb-3](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/520f2cab-9b62-4fce-a8c9-04d5a2e5a868)

* 새롭게 결정된 부모와 자식의 색을 변경
![rb-4](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/8ad4bf0d-ad32-4ac7-b317-1affcc23b61b)

* 위에서 노드 2는 2개의 NIL을 가지고, 이 NIL을 검은색 노드라고 생각한다면 조건 3을 해치지 않게 된다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 레드-블랙트리의 Recoloring에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

1. P와 U를 검은색으로 바꾸고 G를 빨간색으로 바꾼다.
2. G가 루트 노드라면 검은색으로 바꾼다.
3. G를 빨간색으로 바꿨을 때 또다시 Double Red가 발생한다면 또다시 Restructuring 혹은 Recoloring을 진행해서 Double Red 문제가 발생하지 않을 때까지 반복한다.

* Double Red &  U가 빨간색 -> Recoloring 수행
![rb-5](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/61b74b0f-7811-44d0-88f1-5c737fb2aa75)

* U와 G 색깔 변경
   * 이 예시에서는 노드 5가 루트이므로 검은색으로 변경
   * 루트 노드는 검은색이어야 하고, 검은색 노드는 연속해서 나올 수 있으므로 상관 없음
![rb-6](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/6ee42228-5dff-4083-8304-7c57087eedc6)

* Recoloring을 했는데, 또 Double Red가 발생할 수 있음
![rb-7](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/deb0f097-7aac-4b31-a0b2-3f2bf5a89bf6)

* Recoloring 또는 Restructuring을 문제가 해결될 때까지 진행
![rb-8](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/51079971-e602-4de1-832e-1158c4f2692c)

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ AVL트리에 대해 해당 트리의 삽입, 검색, 삭제의 시간 복잡도를 포함하여 설명해주세요.

<br>

### 💬 답변
<blockquote>

![img1](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/caa2fc17-22c8-4b63-800a-6e75e4e669c3)

AVL트리는 이진 탐색 트리가 편향적으로 이루어졌을 때 탐색 속도가 안 좋아지는 것은 개선하기 위해 고안된 트리이다. 위 그림처럼 편향된 이진 트리인 경우, 아래 그림처럼 회전을 통해 높이를 균일하게 맞춰준다.

![img2](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/f16febde-07ae-4d47-8a80-59d615aec319)

따라서 AVL트리는 다음과 같은 조건을 만족한다.

- 이진 탐색 트리의 속성을 가짐
- 왼쪽, 오른쪽 서브 트리의 높이 차가 최대 1
- 높이 차가 1보다 커지면 회전을 통해 균형을 맞춤
- 높이를 일정하게 맞춤으로써 삽입, 삭제, 검색의 시간 복잡도가 O(logN) 보장

</blockquote>

[AVL트리 시뮬레이션 사이트](https://www.cs.usfca.edu/~galles/visualization/AVLtree.html)

<br>

**Comment by seongeun42** [comment]
### ❓️ 트리 순회를 위해 사용하는 방법에는 무엇이 있는지 설명해주세요.

<br>

### 💬 답변
<blockquote>

- **전위 순회(preorder)**
  - 루트 → 왼쪽 → 오른쪽 순으로 순회하는 방식
![preorder](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/51b29c66-f952-4cd3-b247-92908d13f9df)

- **중위 순회(inorder)**
  - 왼쪽 → 루트 → 오른쪽 순으로 순회하는 방식
![inorder](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/c796d3aa-d71e-4d14-9d83-d1272d2c1ad4)

- **후위 순회(postorder)**
  - 왼쪽 → 오른쪽 → 루트 순으로 순회하는 방식
![postorder](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/7807bf4f-da70-4c84-9337-f4b0f15c8b41)

- **층별 순회(level order)**
  - 위에서부터 level별로 순서대로 순회하는 방식
![levelorder](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/2752cc2e-d6da-42ae-a484-cd92d14d726d)

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 주어진 트리의 높이와 차수를 말해주세요.
![tree](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/8e295580-04d5-47f9-b5c0-1b886cfacb5d)

<br>

### 💬 답변
<blockquote>

높이 : 4
차수 : 3

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ BST(Binary Search Tree)와 Binary Tree에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

이진 트리(Binary Tree)란 각각의 노드가 최대 두 개의 자식 노드를 가지는 트리를 말한다. 이진 트리는 이진 탐색 트리와 이진 힙에 사용된다.

![img1](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/0f161082-14a5-4a44-bd21-73e9ab6dc6dd)

이진 탐색 트리란 특정 값을 찾는 것에 특화된 이진 트리를 말한다. 각 노드에는 값이 있고, 노드의 왼쪽은 해당 노드 값보다 작은 값을 가진 노드여야 하고, 오른쪽은 해당 노드의 값보다 큰 값을 가진 노드여야 한다. 트리 내의 서브 트리도 모두 해당 조건을 만족해야 하며, 중복된 키를 허용하지 않는다.

특정 값을 찾고자 할 때, 이진 탐색 트리의 조건을 이용해 현재 노드보다 찾는 값이 작다면 왼쪽, 크다면 오른쪽으로 이동해 이동하는 식으로 찾는다면 검색 속도를 $O(log_2n)$으로 단축할 수 있다.

</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️ 힙 트리(Heap Tree)에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

* 완전 이진 트리의 일종으로, 여러 값 중 가장 크거나 작은 값을 빠르게 찾기 위해 고안
   * 모든 노드가 0개 또는 2개의 자식 노드를 가짐
* Max heap의 경우, 자식 노드의 값은 부모 노드보다 작음
* Min heap의 경우, 자식 노드의 값은 부모 노드보다 큼
![heap](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/912c01ef-2164-4305-89ef-3e31ccccc2e8)

* 우선순위 큐나 힙 소트 구현에 활용됨.
* Min heap의 경우 최솟값을, Max heap의 경우 최댓값을 $O(1)$에 찾을 수 있음.
* 데이터 삽입, 삭제는 $O(logN)$.
   * 데이터 삽입 시, 트리의 마지막에 삽입
      * 완전 이진 트리의 구조를 지키기 위해
   * 이후, 부모 노드와 비교하면서 부모 노드보다 값이 크면 서로 Swap.
      * Max heap의 경우. Min heap의 경우는 반대.
  * 데이터 삭제 시, 루트 노드를 삭제
  * 이후 힙의 마지막 노드를 루트로 옮김
  * 트리 구조 및 힙의 종류에 따라 구조 재조정 

</blockquote>

<br>
